File: packages/cli/src/services/linker.ts
import { execFile } from 'node:child_process';
import { mkdir, symlink, readlink, unlink, readdir, stat, cp } from 'node:fs/promises';
import * as path from 'node:path';
import * as os from 'node:os';
import { logVerbose } from '../lib/output.js';

/** Agent directory definitions */
interface AgentDir {
  name: string;
  skillsDir: string;
}

const getAgentDirs = (): AgentDir[] => {
  const home = os.homedir();
  return [
    { name: 'Claude Code', skillsDir: path.join(home, '.claude', 'skills') },
    { name: 'Cursor', skillsDir: path.join(home, '.cursor', 'skills') },
    { name: 'Codex', skillsDir: path.join(home, '.agents', 'skills') },
  ];
};

interface LinkResult {
  agents: string[];
  method: 'vercel-skills-cli' | 'symlink' | 'copy';
  isCopy: boolean;
}

/**
 * Execute npx skills add with timeout.
 * Returns parsed agent names on success, null on failure.
 */
const tryVercelSkillsCli = (skillPath: string): Promise<string[] | null> => {
  return new Promise((resolve) => {
    execFile(
      'npx',
      ['skills', 'add', skillPath, '-a', '*', '-y'],
      { timeout: 30_000 },
      (error, stdout) => {
        if (error) {
          resolve(null);
          return;
        }
        // Parse output for agent names; each linked agent typically shows on its own line
        const agents: string[] = [];
        const lines = stdout.split('\n');
        for (const line of lines) {
          const lower = line.toLowerCase();
          if (lower.includes('claude')) agents.push('Claude Code');
          if (lower.includes('cursor')) agents.push('Cursor');
          if (lower.includes('codex')) agents.push('Codex');
          if (lower.includes('copilot')) agents.push('Copilot');
        }
        resolve(agents.length > 0 ? agents : ['linked']);
      },
    );
  });
};

/**
 * Try to create a symlink to the agent's skills directory.
 * Returns true on success, false on failure.
 */
const trySymlink = async (skillPath: string, targetDir: string): Promise<boolean> => {
  try {
    await mkdir(path.dirname(targetDir), { recursive: true });

    // Remove existing link/dir if present
    try {
      const existing = await readlink(targetDir);
      if (existing === skillPath) return true; // already linked correctly
      await unlink(targetDir);
    } catch {
      // No existing link, or it's a regular file/dir — try to remove
      try {
        await unlink(targetDir);
      } catch {
        // Nothing to remove
      }
    }

    await symlink(skillPath, targetDir, 'dir');
    return true;
  } catch {
    return false;
  }
};

/**
 * Copy skill files as last resort.
 */
const tryCopy = async (skillPath: string, targetDir: string): Promise<boolean> => {
  try {
    await mkdir(targetDir, { recursive: true });
    await cp(skillPath, targetDir, { recursive: true });
    return true;
  } catch {
    return false;
  }
};

/**
 * Link a skill to all detected agent directories.
 * Fallback chain: Vercel skills CLI -> symlink -> copy
 */
export const linkSkill = async (skillPath: string, skillName: string): Promise<LinkResult> => {
  // Method 1: Try Vercel skills CLI
  logVerbose(`Trying Vercel skills CLI for ${skillName}...`);
  const vercelAgents = await tryVercelSkillsCli(skillPath);
  if (vercelAgents) {
    logVerbose(`Linked via Vercel skills CLI to: ${vercelAgents.join(', ')}`);
    return { agents: vercelAgents, method: 'vercel-skills-cli', isCopy: false };
  }

  // Method 2: Try direct symlinks to agent dirs
  logVerbose('Vercel skills CLI not available, trying symlinks...');
  const agentDirs = getAgentDirs();
  const linkedAgents: string[] = [];

  for (const agent of agentDirs) {
    const targetDir = path.join(agent.skillsDir, skillName);
    const success = await trySymlink(skillPath, targetDir);
    if (success) {
      linkedAgents.push(agent.name);
      logVerbose(`Symlinked to ${agent.name}: ${targetDir}`);
    }
  }

  if (linkedAgents.length > 0) {
    return { agents: linkedAgents, method: 'symlink', isCopy: false };
  }

  // Method 3: Fall back to file copy
  logVerbose('Symlinks failed, falling back to copy...');
  const copiedAgents: string[] = [];

  for (const agent of agentDirs) {
    const targetDir = path.join(agent.skillsDir, skillName);
    const success = await tryCopy(skillPath, targetDir);
    if (success) {
      copiedAgents.push(agent.name);
      logVerbose(`Copied to ${agent.name}: ${targetDir}`);
    }
  }

  return { agents: copiedAgents, method: 'copy', isCopy: true };
};

/**
 * Unlink a skill from all agent directories.
 * Always does manual removal to ensure symlinks/copies are cleaned up,
 * and optionally tries Vercel skills CLI for additional cleanup.
 */
export const unlinkSkill = async (skillName: string): Promise<string[]> => {
  const agentDirs = getAgentDirs();
  const unlinkedAgents: string[] = [];

  // Manual removal from all agent dirs
  for (const agent of agentDirs) {
    const targetDir = path.join(agent.skillsDir, skillName);
    try {
      // Use lstat to check without following symlinks
      await stat(targetDir).catch(() => null);
      const linkTarget = await readlink(targetDir).catch(() => null);

      if (linkTarget !== null) {
        // It's a symlink — remove it
        await unlink(targetDir);
        unlinkedAgents.push(agent.name);
      } else {
        // Check if it's a directory (copy)
        const s = await stat(targetDir).catch(() => null);
        if (s?.isDirectory()) {
          const { rm } = await import('node:fs/promises');
          await rm(targetDir, { recursive: true, force: true });
          unlinkedAgents.push(agent.name);
        }
      }
    } catch {
      // Directory doesn't exist or can't be accessed, skip
    }
  }

  // Also try Vercel skills CLI for any agents we might have missed
  try {
    await new Promise<void>((resolve, reject) => {
      execFile('npx', ['skills', 'remove', skillName, '-y'], { timeout: 30_000 }, (error) => {
        if (error) reject(error);
        else resolve();
      });
    });
  } catch {
    // Vercel skills CLI not available or failed — that's fine
  }

  return unlinkedAgents;
};

/**
 * Get list of agents that have a given skill linked.
 */
export const getLinkedAgents = async (skillName: string): Promise<string[]> => {
  const agentDirs = getAgentDirs();
  const linked: string[] = [];

  for (const agent of agentDirs) {
    const targetDir = path.join(agent.skillsDir, skillName);
    try {
      await stat(targetDir);
      linked.push(agent.name);
    } catch {
      // Not present
    }
  }

  return linked;
};

/**
 * Check if a skill was installed via copy (not symlink).
 */
export const isSkillCopy = async (skillName: string): Promise<boolean> => {
  const agentDirs = getAgentDirs();
  for (const agent of agentDirs) {
    const targetDir = path.join(agent.skillsDir, skillName);
    try {
      await readlink(targetDir);
      return false; // It's a symlink
    } catch {
      try {
        await stat(targetDir);
        return true; // Exists but not a symlink — it's a copy
      } catch {
        // Doesn't exist in this agent dir
      }
    }
  }
  return false;
};

/**
 * List all skills found in a specific agent's skills directory.
 */
export const listAgentSkills = async (agentIndex: number): Promise<string[]> => {
  const agentDirs = getAgentDirs();
  if (agentIndex < 0 || agentIndex >= agentDirs.length) return [];

  const agent = agentDirs[agentIndex];
  try {
    const entries = await readdir(agent.skillsDir);
    return entries;
  } catch {
    return [];
  }
};

File: cli-go/internal/linker/linker.go
// Package linker manages linking skills to agent directories.
// It creates symlinks (preferred) or copies files (fallback) into
// agent-specific skill directories like ~/.claude/skills/, ~/.cursor/skills/, etc.
package linker

import (
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

// AgentDir represents a known agent's skills directory.
type AgentDir struct {
	Name      string
	SkillsDir string
}

// LinkResult contains the result of a link operation.
type LinkResult struct {
	Agents []string
	Method string // "symlink" or "copy"
	IsCopy bool
}

// DefaultAgentDirs returns the default agent directories based on the home directory.
func DefaultAgentDirs(homeDir string) []AgentDir {
	return []AgentDir{
		{Name: "Claude Code", SkillsDir: filepath.Join(homeDir, ".claude", "skills")},
		{Name: "Cursor", SkillsDir: filepath.Join(homeDir, ".cursor", "skills")},
		{Name: "Codex", SkillsDir: filepath.Join(homeDir, ".agents", "skills")},
	}
}

// Linker manages skill-to-agent directory linking.
type Linker struct {
	AgentDirs []AgentDir
}

// New creates a new Linker with the given agent directories.
func New(agentDirs []AgentDir) *Linker {
	return &Linker{AgentDirs: agentDirs}
}

// NewDefault creates a new Linker with default agent directories.
func NewDefault(homeDir string) *Linker {
	return New(DefaultAgentDirs(homeDir))
}

// bareName extracts the short name from a scoped skill name.
// "@scope/my-skill" → "my-skill", "my-skill" → "my-skill"
func bareName(name string) string {
	if i := strings.LastIndex(name, "/"); i >= 0 {
		return name[i+1:]
	}
	return name
}

// LinkSkill links a skill to all agent directories.
// Tries symlink first, falls back to copy if symlink fails.
// Scoped names like "@scope/my-skill" are linked as just "my-skill"
// so the slash command is /my-skill instead of /@scope/my-skill.
func (l *Linker) LinkSkill(skillPath string, skillName string) (*LinkResult, error) {
	linkedAgents := make([]string, 0)
	dirName := bareName(skillName)

	// Try symlinks first
	for _, agent := range l.AgentDirs {
		targetDir := filepath.Join(agent.SkillsDir, dirName)
		if err := trySymlink(skillPath, targetDir); err == nil {
			linkedAgents = append(linkedAgents, agent.Name)
		}
	}

	if len(linkedAgents) > 0 {
		return &LinkResult{
			Agents: linkedAgents,
			Method: "symlink",
			IsCopy: false,
		}, nil
	}

	// Fallback to copy
	copiedAgents := make([]string, 0)
	for _, agent := range l.AgentDirs {
		targetDir := filepath.Join(agent.SkillsDir, dirName)
		if err := tryCopy(skillPath, targetDir); err == nil {
			copiedAgents = append(copiedAgents, agent.Name)
		}
	}

	return &LinkResult{
		Agents: copiedAgents,
		Method: "copy",
		IsCopy: true,
	}, nil
}

// UnlinkSkill removes a skill from all agent directories.
// Returns the list of agent names from which the skill was removed.
func (l *Linker) UnlinkSkill(skillName string) ([]string, error) {
	unlinkedAgents := make([]string, 0)
	dirName := bareName(skillName)

	for _, agent := range l.AgentDirs {
		targetDir := filepath.Join(agent.SkillsDir, dirName)

		// Check if it's a symlink
		fi, err := os.Lstat(targetDir)
		if err != nil {
			continue // doesn't exist, skip
		}

		if fi.Mode()&os.ModeSymlink != 0 {
			// It's a symlink — remove it
			if err := os.Remove(targetDir); err == nil {
				unlinkedAgents = append(unlinkedAgents, agent.Name)
			}
		} else if fi.IsDir() {
			// It's a copy — remove recursively
			if err := os.RemoveAll(targetDir); err == nil {
				unlinkedAgents = append(unlinkedAgents, agent.Name)
			}
		}
	}

	return unlinkedAgents, nil
}

// GetLinkedAgents returns the list of agent names where the skill is linked.
func (l *Linker) GetLinkedAgents(skillName string) []string {
	linked := make([]string, 0)
	dirName := bareName(skillName)

	for _, agent := range l.AgentDirs {
		targetDir := filepath.Join(agent.SkillsDir, dirName)
		if _, err := os.Lstat(targetDir); err == nil {
			linked = append(linked, agent.Name)
		}
	}

	return linked
}

// IsSkillCopy returns true if the skill in agent directories is a copy (not a symlink).
func (l *Linker) IsSkillCopy(skillName string) bool {
	dirName := bareName(skillName)
	for _, agent := range l.AgentDirs {
		targetDir := filepath.Join(agent.SkillsDir, dirName)
		fi, err := os.Lstat(targetDir)
		if err != nil {
			continue
		}
		if fi.Mode()&os.ModeSymlink != 0 {
			return false // found a symlink
		}
		if fi.IsDir() {
			return true // found a copy
		}
	}
	return false
}

// trySymlink creates a symlink at targetDir pointing to skillPath.
// Creates parent directories as needed. Removes existing link/dir if present.
func trySymlink(skillPath string, targetDir string) error {
	// Ensure parent directory exists
	if err := os.MkdirAll(filepath.Dir(targetDir), 0o755); err != nil {
		return fmt.Errorf("creating parent directory: %w", err)
	}

	// Check for existing entry
	fi, err := os.Lstat(targetDir)
	if err == nil {
		if fi.Mode()&os.ModeSymlink != 0 {
			// Read existing symlink target
			existing, err := os.Readlink(targetDir)
			if err == nil && existing == skillPath {
				return nil // already linked correctly
			}
			// Remove stale symlink
			if err := os.Remove(targetDir); err != nil {
				return fmt.Errorf("removing existing symlink: %w", err)
			}
		} else {
			// Remove existing file/dir
			if err := os.RemoveAll(targetDir); err != nil {
				return fmt.Errorf("removing existing entry: %w", err)
			}
		}
	}

	return os.Symlink(skillPath, targetDir)
}

// tryCopy copies the skill directory to targetDir.
func tryCopy(skillPath string, targetDir string) error {
	if err := os.MkdirAll(targetDir, 0o755); err != nil {
		return fmt.Errorf("creating target directory: %w", err)
	}

	return filepath.WalkDir(skillPath, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		relPath, err := filepath.Rel(skillPath, path)
		if err != nil {
			return err
		}

		targetPath := filepath.Join(targetDir, relPath)

		if d.IsDir() {
			return os.MkdirAll(targetPath, 0o755)
		}

		return copyFile(path, targetPath)
	})
}

// copyFile copies a single file from src to dst.
func copyFile(src, dst string) error {
	srcFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer func() { _ = srcFile.Close() }()

	dstFile, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer func() { _ = dstFile.Close() }()

	_, err = io.Copy(dstFile, srcFile)
	return err
}
